### ISSUE TITLE
About LR scheduler

### ISSUE THREAD

**Issue by morestart** [Lightning-AI/lightning#19073]
### Description & Motivation

In the older versions, one could use from transformers import get_linear_schedule_with_warmup, get_constant_schedule_with_warmup, get_cosine_schedule_with_warmup. However, in the new versions, the same import statement results in an error:
```
lightning_lite.utilities.exceptions.MisconfigurationException: The provided lr scheduler `LambdaLR` doesn't follow PyTorch's LRScheduler API. You should override the `LightningModule.lr_scheduler_step` hook with your own logic if you are using a custom LR scheduler.
```

this is my code:
```python

@property
    def num_training_steps(self) -> int:
        return self.trainer.estimated_stepping_batches

    def compute_warmup(self, num_training_steps: int, num_warmup_steps: Union[int, float]) -> Tuple[int, int]:
        if num_training_steps < 0:
            # less than 0 specifies to infer number of training steps
            num_training_steps = self.num_training_steps
        if isinstance(num_warmup_steps, float):
            # Convert float values to percentage of training steps to use as warmup
            num_warmup_steps *= num_training_steps
        return num_training_steps, num_warmup_steps

    def get_scheduler(self, scheduler_type, optimizer, num_train_steps, num_warmup_steps):
        if scheduler_type == 'linear':
            scheduler = get_linear_schedule_with_warmup(
                optimizer, num_warmup_steps=num_warmup_steps, num_training_steps=num_train_steps
            )
            return scheduler
        elif scheduler_type == 'constant':
            scheduler = get_constant_schedule_with_warmup(
                optimizer, num_warmup_steps=num_warmup_steps
            )
            return scheduler
        elif scheduler_type == 'cosine':
            scheduler = get_cosine_schedule_with_warmup(
                optimizer, num_warmup_steps=num_warmup_steps, num_training_steps=num_train_steps,
                num_cycles=self.num_cycles
            )
            return scheduler
        else:
            raise NotImplementedError('Scheduler not implemented')

    def configure_optimizers(self):
        # return {"optimizer": optimizer, "lr_scheduler": scheduler}
        num_training_steps, num_warmup_steps = self.compute_warmup(
            num_training_steps=-1,
            num_warmup_steps=0.2,
        )
        lr_scheduler = self.get_scheduler(self.scheduler_type, optimizer, num_training_steps, num_warmup_steps)
        scheduler = {"scheduler": lr_scheduler, "interval": "step", "frequency": 1}
        return {"optimizer": optimizer, "lr_scheduler": scheduler}
```

```

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

cc @borda

**Comment by awaelchli** [comment]
> However, in the new versions, the same import statement results in an error:

In which version? I fail to reproduce this. Here is the code that you provided embedded in a full example and it works for every scheduler type:

```py
from typing import Union, Tuple

import torch
from transformers import (
    get_linear_schedule_with_warmup,
    get_constant_schedule_with_warmup,
    get_cosine_schedule_with_warmup
)
from lightning.pytorch import LightningModule, Trainer
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.scheduler_type = "constant"

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    @property
    def num_training_steps(self) -> int:
        return self.trainer.estimated_stepping_batches

    def compute_warmup(self, num_training_steps: int, num_warmup_steps: Union[int, float]) -> Tuple[int, int]:
        if num_training_steps < 0:
            # less than 0 specifies to infer number of training steps
            num_training_steps = self.num_training_steps
        if isinstance(num_warmup_steps, float):
            # Convert float values to percentage of training steps to use as warmup
            num_warmup_steps *= num_training_steps
        return num_training_steps, num_warmup_steps

    def get_scheduler(self, scheduler_type, optimizer, num_train_steps, num_warmup_steps):
        if scheduler_type == 'linear':
            scheduler = get_linear_schedule_with_warmup(
                optimizer, num_warmup_steps=num_warmup_steps, num_training_steps=num_train_steps
            )
            return scheduler
        elif scheduler_type == 'constant':
            scheduler = get_constant_schedule_with_warmup(
                optimizer, num_warmup_steps=num_warmup_steps
            )
            return scheduler
        elif scheduler_type == 'cosine':
            scheduler = get_cosine_schedule_with_warmup(
                optimizer, num_warmup_steps=num_warmup_steps, num_training_steps=num_train_steps,
                num_cycles=5,
            )
            return scheduler
        else:
            raise NotImplementedError('Scheduler not implemented')

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        num_training_steps, num_warmup_steps = self.compute_warmup(
            num_training_steps=-1,
            num_warmup_steps=0.2,
        )
        lr_scheduler = self.get_scheduler(self.scheduler_type, optimizer, num_training_steps, num_warmup_steps)
        scheduler = {"scheduler": lr_scheduler, "interval": "step", "frequency": 1}
        return {"optimizer": optimizer, "lr_scheduler": scheduler}


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(max_epochs=2)
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)


if __name__ == "__main__":
    run()

```

Using transformers==4.35.2, ran with Lightning 2.0.7, 2.1.2, and master.
Can you please check again?

**Comment by morestart** [comment]
>>> import lightning
>>> lightning.__version__
'2022.10.25'
>>>

**Comment by morestart** [comment]
>>> import lightning
>>> lightning.__version__
'2022.10.25'
>>>

**Comment by awaelchli** [comment]
@morestart Thanks, that's quite old. It looks like this is fixed though with the latest versions of Lightning and PyTorch. I suggest that you upgrade the version.

**Comment by morestart** [comment]
> @morestart Thanks, that's quite old. It looks like this is fixed though with the latest versions of Lightning and PyTorch. I suggest that you upgrade the version.
I installed lightning via pip. Could there possibly be a conflict with my environment? My PyTorch version is 2.1.

**Comment by awaelchli** [comment]
Yes, possibly, because 2022.10.25 is actually an inofficial version that was removed from pypi. You could try to force install it like so: `pip install "lightning>=2.0"`.

**Comment by morestart** [comment]
Thank you very much for your reply!
